File: aerial-control-lab/python/exercises/Week4/Mon/task_B/rk4_oscillator_rewrite.py
import numpy as np
from aerial_control.integrators import rk4_step

def oscillator_derivative(t, state):
    return np.array([state[1], -state[0]])
    # state 的顺序是 [q, v]
    # 返回状态导数 [v, -q]

def rk4_step_independent(derivative, t, state, dt):
    k1 = derivative(t, state)
    k2 = derivative(t + dt / 2, state + dt / 2 * k1)
    k3 = derivative(t + dt / 2, state + dt / 2 * k2)
    k4 = derivative(t + dt, state + dt * k3)
    # 依次计算 k1、k2、k3、k4
    # 返回经过一个 dt 后的新状态
    return state + dt / 6 * (k1 + 2 * k2 + 2 * k3 + k4)

def simulate(step_function):
    initial_state = np.array([1.0, 0.0])  # 初始状态 [q, v]
    dt = 0.1  # 时间步长
    t = 0.0  # 初始时间
    num_steps = 10
    for _ in range(num_steps):  # 使用下划线做丢弃的循环变量，不要复用step_function
        initial_state = step_function(oscillator_derivative, t, initial_state, dt)
        t += dt
    return initial_state  # return挪到循环外面

analytical = np.array([np.cos(1.0), -np.sin(1.0)])


independent_result = simulate(rk4_step_independent)
official_result = simulate(rk4_step)
analytical_result = np.array([np.cos(1.0), -np.sin(1.0)])
print("Independent RK4 result:", independent_result)
print("Official RK4 result:", official_result)
print("Analytical result:", analytical_result)
print(np.max(np.abs(independent_result - analytical)))
print(np.max(np.abs(independent_result - official_result)))

File: aerial-control-lab/cpp/exercises/Week4/Mon/task_B/rk4_oscillator_rewrite.cpp

#include "core/integrators.hpp"

#include <cmath>
#include <functional>
#include <iomanip>
#include <iostream>

// 调用链：
// main
//   -> simulate(rk4StepIndependent 或 aerial_control::rk4Step)
//      -> 单步函数
//         -> oscillatorDerivative（RK4 每一步调用四次）

using StepFunction = std::function<aerial_control::StateVector(
    const aerial_control::DerivativeFunction&,
    double,
    const aerial_control::StateVector&,
    double)>;

constexpr double kInitialTime = 0.0;
constexpr double kTimeStep = 0.1;
constexpr int kStepCount = 10;
constexpr double kEndTime = kInitialTime + kTimeStep * kStepCount;

// 第一层：具体动力学。输入状态 [q, v]，返回状态导数 [v, -q]。
aerial_control::StateVector oscillatorDerivative(
    double /*t*/, const aerial_control::StateVector& state) {
  aerial_control::StateVector result(2);
  result << state(1), -state(0);
  return result;
}

// 第二层：通用 RK4 单步。它通过 derivative 获取动力学，不了解简谐振子细节。
aerial_control::StateVector rk4StepIndependent(
    const aerial_control::DerivativeFunction& derivative,
    double t,
    const aerial_control::StateVector& state,
    double dt) {
  const aerial_control::StateVector k1 = derivative(t, state);
  const aerial_control::StateVector k2 =
      derivative(t + dt / 2, state + dt / 2 * k1);
  const aerial_control::StateVector k3 =
      derivative(t + dt / 2, state + dt / 2 * k2);
  const aerial_control::StateVector k4 =
      derivative(t + dt, state + dt * k3);

  return state + dt / 6 * (k1 + 2 * k2 + 2 * k3 + k4);
}

// 第三层：仿真循环。它只负责重复调用传入的单步函数并推进时间。
aerial_control::StateVector simulate(const StepFunction& stepFunction) {
  aerial_control::StateVector state(2);
  state << 1.0, 0.0;

  double t = kInitialTime;

  for (int i = 0; i < kStepCount; ++i) {
    state = stepFunction(oscillatorDerivative, t, state, kTimeStep);
    t += kTimeStep;
  }

  return state;
}
 
// 第四层：程序入口。选择两种单步函数，运行相同仿真并比较结果。
int main() {
  const aerial_control::StateVector independentResult =
      simulate(rk4StepIndependent);
  const aerial_control::StateVector officialResult =
      simulate(aerial_control::rk4Step);

  aerial_control::StateVector analyticalResult(2);
  analyticalResult << std::cos(kEndTime), -std::sin(kEndTime);

  const double independentAnalyticalError =
      (independentResult - analyticalResult).cwiseAbs().maxCoeff();
  const double independentOfficialDifference =
      (independentResult - officialResult).cwiseAbs().maxCoeff();

  std::cout << std::setprecision(17);
  std::cout << "Independent RK4 result: " << independentResult.transpose()
            << '\n';
  std::cout << "Official RK4 result: " << officialResult.transpose() << '\n';
  std::cout << "Analytical result: " << analyticalResult.transpose() << '\n';
  std::cout << "Independent vs analytical max error: "
            << independentAnalyticalError << '\n';
  std::cout << "Independent vs official max difference: "
            << independentOfficialDifference << '\n';

  return 0;
}
